crates/spirv-builder/build.rs

```diff
@@ -0,0 +1,233 @@
+    if propagate_features.is_empty() {
+        return;
+    }
+
+    // Clear Cargo environment variables that we don't want to leak into the
+    // inner invocations of Cargo (because e.g. build scripts might read them).
+    for (key, _) in env::vars_os() {
+        let remove = key.to_str().map_or(false, |s| {
+            s.starts_with("CARGO_FEATURES_") || s.starts_with("CARGO_CFG_")
+        });
+        if remove {
+            env::remove_var(key);
+        }
+    }
+
+    let profile = env::var("PROFILE").unwrap();
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+
+    // Try to obtain the outer target dir, to base the target dir we'll build
+    // `rustc_codegen_spirv` in, off of it (but we can use `$OUT_DIR` as fallback).
+    let target_dir_base = {
+        // Strip `$outer_profile/build/*/out`.
+        let outer_target_dir = [&profile, "build", "*", "out"].iter().rev().try_fold(
+            out_dir.clone(),
+            |mut dir, &filter| {
+                if (filter == "*" || dir.ends_with(filter)) && dir.pop() {
+                    Some(dir)
+                } else {
+                    None
+                }
+            },
+        );
+        match outer_target_dir {
+            Some(target) => target.join("rustc_codegen_spirv-build"),
+            None => out_dir.join("build"),
+        }
+    };
+
+    // Set up a workspace in `$OUT_DIR` to bootstrap building this very crate
+    // (`spirv-builder`) with `rust-toolchain.toml` and `Cargo.lock` taken from
+    // our `rustc_codegen_spirv` dependency.
+    let ws = out_dir.join("nested-self-builder");
+    fs::create_dir_all(ws.join("src")).unwrap();
+    fs::write(
+        ws.join("Cargo.toml"),
+        format!(
+            r#"
+[workspace]
+
+[package]
+name = "nested-self-builder"
+edition = "2021"
+version = "0.0.0"
+
+[dependencies.spirv-builder]
+path = {spirv_builder_path:?}
+default-features = false
+features = [{spirv_builder_features}]
+
+# Enable incremental by default in release mode, as well.
+[profile.release]
+incremental = true
+"#,
+            spirv_builder_path = env!("CARGO_MANIFEST_DIR"),
+            spirv_builder_features = propagate_features
+                .iter()
+                .copied()
+                .chain(["internal-unstable-trigger-self-build-for-backend"])
+                .map(|feat| format!("{feat:?}"))
+                .collect::<Vec<_>>()
+                .join(", "),
+        ),
+    )
+    .unwrap();
+    const NESTED_SELF_BUILDER_MAIN_RS: &str = r#"
+fn main() {
+    let dylib_path = spirv_builder::codegen_backend::codegen_backend_dylib_path();
+    println!("cargo:rerun-if-changed={}", dylib_path.display());
+    println!("cargo:rustc-env=SPIRV_BUILDER_CODEGEN_BACKEND_DYLIB_PATH={}", dylib_path.display());
+    println!("cargo:rustc-env=SPIRV_BUILDER_RUSTUP_TOOLCHAIN={}", env!("RUSTUP_TOOLCHAIN"));
+}
+"#;
+    let ws_src_main_rs = ws.join("src/main.rs");
```

Comment:
New idea (from https://github.com/EmbarkStudios/rust-gpu/issues/911#issuecomment-1966153512) to consider trying:

The diff I'm commenting on _generates a pair of `Cargo.toml` and `src/main.rs`_ - it could be a `-Z script` combined `.rs` file, _forcing Cargo to help out with caching_.

That is, we could manage `~/.cache/rust-gpu` (something respecting `$XDG_CACHE_HOME` in reality), under which we generate a `-Z script` `.rs` file per _choice of contents_ (as in, we can content-address it, i.e. the files just take a name given by their own hash).

That directory of `.rs` files would be relatively miniscule, but Cargo should be able to e.g. automatically GC the target dirs _it created itself (under `~/.cargo`) to build `-Z script` packages_.